artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/sasl/MechanismFinder.java

```diff
@@ -17,11 +17,41 @@
 
 package org.apache.activemq.artemis.protocol.amqp.sasl;
 
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.ServiceLoader;
+
 public class MechanismFinder {
 
-   public static String[] KNOWN_MECHANISMS = new String[]{PlainSASL.NAME, AnonymousServerSASL.NAME};
+   private static final Map<String, ServerSASLFactory> FACTORY_MAP = new HashMap<>();
+   private static final Comparator<? super ServerSASLFactory> PRECEDENCE_COMPARATOR =
+      (f1, f2) -> Integer.compare(f1.getPrecedence(), f2.getPrecedence());
+
+   static {
+      ServiceLoader<ServerSASLFactory> serviceLoader =
+               ServiceLoader.load(ServerSASLFactory.class, MechanismFinder.class.getClassLoader());
+      for (ServerSASLFactory factory : serviceLoader) {
+         FACTORY_MAP.merge(factory.getMechanism(), factory, (f1, f2) -> {
+            if (f2.getPrecedence() > f1.getPrecedence()) {
+               return f2;
+            } else {
+               return f1;
+            }
+         });
+      }
+   }
+
+   public static String[] getDefaultMechanisms() {
+      return FACTORY_MAP.values()
+            .stream()
+            .filter(ServerSASLFactory::isDefaultPermitted)
+            .sorted(PRECEDENCE_COMPARATOR.reversed())
+            .map(ServerSASLFactory::getMechanism)
+            .toArray(String[]::new);
```

Comment:
This is still being evaluated on every get, computing the same result over and over for every connection. Determine the result while creating the map and this can then simply return it.

Reply:
The value is just read once on connector creation if I see right, nerveless I'll extract this so it is only computed once.